llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -2147,14 +2148,57 @@ static SDValue foldAddSubBoolOfMaskedVal(SDNode *N, SelectionDAG &DAG) {
   return DAG.getNode(IsAdd ? ISD::SUB : ISD::ADD, DL, VT, C1, LowBit);
 }
 
+/// Try to fold a pointer arithmetic node, preferring integer arithmetic.
+/// This needs to be done separately from normal addition, because pointer
+/// addition is not commutative.
 SDValue DAGCombiner::visitPTRADD(SDNode *N) {
   SDValue N0 = N->getOperand(0);
   SDValue N1 = N->getOperand(1);
+  EVT PtrVT = N0.getValueType();
+  EVT IntVT = N1.getValueType();
+  SDLoc DL(N);
+
+  // fold (ptradd undef, y) -> undef
+  if (N0.isUndef())
+    return N0;
+
+  // fold (ptradd x, undef) -> undef
+  if (N1.isUndef())
+    return DAG.getUNDEF(PtrVT);
+
   // fold (ptradd x, 0) -> 0
-  // This needs to be done separately from normal addition, because pointer
-  // addition is not commutative.
   if (isNullConstant(N1))
     return N0;
+
+  // Reassociate: (ptradd (ptradd x, y), z) -> (ptradd x, (add y, z)) if:
+  //   * x is a null pointer; or
+  //   * the add can be constant-folded; or
+  //   * the add can be combined and z is not a constant; or
+  //   * y is a constant and z has one use; or
+  //   * y is a constant and (ptradd x, y) has one use; or
+  //   * (ptradd x, y) and z have one use and z is not a constant.
```

Comment:
That and because I want to keep a constant on the outer ptradd so it can easily treated as an immediate; would be concerned about obfuscating it too much for patterns (although `(add expr, const)` is the canonical form for adds so maybe that's less of a concern).